### PR TITLE
RocksDb Directory Peer Repository

### PR DESCRIPTION
**Pull request by MendelMonteiro**
Implement a peer repository that uses RocksDb as a persistent storage.

### CODE REVIEW

File: src/Abc.Zebus.Directory.RocksDb/Storage/RocksDbPeerRepository.cs [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+        private void IterateDynamicSubscriptionsForPeer(PeerId peerId, Action<byte[], byte[]> action)
+        {
+            var key = BuildSubscriptionKeyPrefix(peerId);
+            using var cursor = _db.NewIterator(_subscriptionsColumnFamily);
+            if (!cursor.Seek(key).Valid())
+                return;
+
+            var peerIdLength = peerId.ToString().Length;
+            byte[] currentKey;
+            do
```
Comment (ltrzesniewski) [inline src/Abc.Zebus.Directory.RocksDb/Storage/RocksDbPeerRepository.cs]: The prefix does not seem to be verified for the first key... are we sure the behavior will be correct if we have two peer ids with a common prefix? I mean for instance if you're looking for `Foo` but the db contains `FooBar` you could get the `FooBar` value.

Likewise, `CompareStart` does not seem to handle this case.

Maybe the key structure should be `PeerId | \0 | MessageFullTypeName` so we could use `PeerId | \0` as the prefix to test?

File: src/Abc.Zebus.Directory.RocksDb/Storage/RocksDbPeerRepository.cs [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+                                                   if (messageTypeIds.Contains(messageTypeId))
+                                                       _db.Remove(key, _subscriptionsColumnFamily);
+                                               });
+        }
+
+        public void RemovePeer(PeerId peerId)
+        {
+            _db.Remove(BuildPeerKey(peerId), _peersColumnFamily);
+
+            IterateDynamicSubscriptionsForPeer(peerId, (key, value) => _db.Remove(key, _subscriptionsColumnFamily));
```
Comment (ltrzesniewski) [inline src/Abc.Zebus.Directory.RocksDb/Storage/RocksDbPeerRepository.cs]: `RemoveAllDynamicSubscriptionsForPeer`

File: src/Abc.Zebus.Directory.RocksDb/Storage/RocksDbPeerRepository.cs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+                return;
+
+            var peer = Get(peerId);
+            if (peer?.TimestampUtc > timestampUtc)
+                return;
+
+            foreach (var subscription in subscriptionsForTypes)
+            {
+                var key = BuildSubscriptionKey(peerId, subscription.MessageTypeId);
+                var value = SerializeBindingKeys(subscription.BindingKeys);
```
Comment (ltrzesniewski) [inline src/Abc.Zebus.Directory.RocksDb/Storage/RocksDbPeerRepository.cs]: Maybe we could use a single `MemoryStream` for all binding keys

File: src/Abc.Zebus.Tests/Abc.Zebus.Tests.csproj [hunk 4]
```diff
@@ -3,6 +3,8 @@
     <TargetFrameworks>net471;netcoreapp2.1</TargetFrameworks>
     <Version>$(ZebusVersion)</Version>
     <IsTestProject>true</IsTestProject>
+    <OutputType>Exe</OutputType>
+    <StartupObject>Abc.Zebus.Tests.Program</StartupObject>
```
Comment (ltrzesniewski) [inline src/Abc.Zebus.Tests/Abc.Zebus.Tests.csproj]: That will prevent tests from running on .NET Core, as the runner defines its own entry point.

File: src/Abc.Zebus.Tests/Program.cs [hunk 5]
```diff
@@ -12,9 +12,9 @@
 
 namespace Abc.Zebus.Tests
 {
-    public static class Program
+    public class Program
     {
-        public static void MainLol(string[] args)
+        public static void Main(string[] args)
```
Comment (ltrzesniewski) [inline src/Abc.Zebus.Tests/Program.cs]: That will prevent tests from running on .NET Core, as the runner defines its own entry point.

File: src/Abc.Zebus.Directory.RocksDb/Storage/RocksDbPeerRepository.cs (hunk 1 above)
Comment (MendelMonteiro) [reply src/Abc.Zebus.Directory.RocksDb/Storage/RocksDbPeerRepository.cs]: You're right we either need to add a separator or check the peer id length encoded in the key.

**Pull request merged**